_posts/2020-04-23-spring-boot-testcontainers.md

```diff
@@ -0,0 +1,215 @@
+
+## Using Production-Like Environment for Migration Tests 
+There is another approach for testing database migration. First, we can test migration with `H2` database at build time.
+After that, we can deploy our application in a production-like environment and let the migration run on this environment with 
+the database like in production, for example, `PostgreSQL`.
+
+This approach has a couple of drawbacks:
+
+* Bugs are discovered too late,
+* it is hard to find errors.
+
+Let's imagine, we test the migration with `H2` database during building the application, and the tests are green.
+The next step is delivering and deploying the application to a test environment. It takes time. If the migration on the test environment fails,
+we know it too late, maybe several minutes later. It slows down the development cycle.
+
+Also, this situation is very confusing for developers, because we can't debug this error with our unit test.
+Our unit test with `H2` was green, and the error occurred on the test environment.
+
+## Solution with Testcontainers
+With [Tescontainers](https://www.testcontainers.org/) we can test the database migration with [Docker](https://www.docker.com/) from our code.
+`Tescontainers` is a Java library, that provides unit testing with Docker.
+Of course, we have to install Docker wherever we want to build the application. After that we can create an initialization code for testing:
+
+````java
+@ContextConfiguration(initializers 
+                               = AbstractIntegrationTest.Initializer.class)
+public class AbstractIntegrationTest {
+
+    static class Initializer
+            implements ApplicationContextInitializer
+                        <ConfigurableApplicationContext> {
+
+        static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>();
+
+        public static Map<String, String> getProperties() {
+            Startables.deepStart(Stream.of(postgres)).join();
+            // we can add further containers here like rabbitmq
+            //or other database
+
+            return Map.of(
+                    "spring.datasource.url", postgres.getJdbcUrl(),
+                    "spring.datasource.username", postgres.getUsername(),
+                    "spring.datasource.password", postgres.getPassword()
+            );
+        }
+
+
+        @Override
+        public void initialize(ConfigurableApplicationContext
+                                           applicationContext) {
+            ConfigurableEnvironment environment 
+                                = applicationContext.getEnvironment();
+            MapPropertySource testcontainers = new MapPropertySource(
+                    "testcontainers",
+                    (Map) getProperties()
+            );
+            environment.getPropertySources().addFirst(testcontainers);
+        }
+    }
+}
+```` 
+The class `AbstractIntegrationTest` is an abstract class, that defines a `PostgreSQL` database and configures 
```

Comment:
The section starting here explaining the code above is a little confusing. One can understand it only after having read through everything. It jumps back and forth between the code lines.

Can you try explaining the code in order. For example:
1. what does the `initializers` in the annotation do? it calls `initialize()`.
2. explain `initialize()`
3. explain `deepStart()` 
4. explain how we pass the connection properties to Spring

Reply:
I refactored the code and change the explanation.